ddtrace/_trace/processor/__init__.py

```diff
@@ -135,30 +135,55 @@ def unregister(self):
-    are unsampled then ``None`` is returned.
+    """Processor that runs both trace and span sampling rules.
 
-    Note that this processor is only effective if complete traces are sent. If
-    the spans of a trace are divided in separate lists then it's possible that
-    parts of the trace are unsampled when the whole trace should be sampled.
+    * Span sampling must be applied after trace sampling priority has been set.
+    * Span sampling rules are specified with a sample rate or rate limit as well as glob patterns
+      for matching spans on service and name.
+    * If the span sampling decision is to keep the span, then span sampling metrics are added to the span.
+    * If a dropped trace includes a span that had been kept by a span sampling rule, then the span is sent to the
+      Agent even if the dropped trace is not (as is the case when trace stats computation is enabled).
     """
 
     _compute_stats_enabled = attr.ib(type=bool)
+    sampler = attr.ib()
+    single_span_rules = attr.ib(type=List[SpanSamplingRule])
 
     def process_trace(self, trace):
         # type: (List[Span]) -> Optional[List[Span]]
+
         if trace:
+            chunk_root = trace[0]
+            root_ctx = chunk_root._context
+
+            # only trace sample if we haven't already sampled
+            if root_ctx and root_ctx.sampling_priority is None:
+                self.sampler.sample(trace[0])
             # When stats computation is enabled in the tracer then we can
             # safely drop the traces.
             if self._compute_stats_enabled:
-                priority = trace[0]._context.sampling_priority if trace[0]._context is not None else None
+                priority = root_ctx.sampling_priority if root_ctx is not None else None
                 if priority is not None and priority <= 0:
                     # When any span is marked as keep by a single span sampling
                     # decision then we still send all and only those spans.
                     single_spans = [_ for _ in trace if is_single_span_sampled(_)]
 
                     return single_spans or None
 
+            # single span sampling rules are applied after trace sampling
+            if self.single_span_rules:
+                for span in trace:
+                    if span.context.sampling_priority is not None and span.context.sampling_priority <= 0:
+                        for rule in self.single_span_rules:
+                            if rule.match(span):
+                                rule.sample(span)
+                                # If stats computation is enabled, we won't send all spans to the agent.
+                                # In order to ensure that the agent does not update priority sampling rates
+                                # due to single spans sampling, we set all of these spans to manual keep.
+                                if config._trace_compute_stats:
```

Comment:
Wait a minute, so if single span sampling rules are set and stats are computed all spans are set to manual keep? Wouldn't this override sampling decisions where we should drop a span? This seems like a bug

Reply:
Nah, this is the desired behavior and is consistent with what it was before. Basically, if we set the single span sampling tags, the trace-agent will take the span, despite the span having `_sampling_prioirity_v1=-1`. Which is fine, however, in the case that we don't send the entire trace (`config._trace_compute_stats=True`), the agent would update priority sampling rates incorrectly, therefore to avoid that, we set all single span sampled spans to manual_keep, so they don't count towards the agent's trace level sample rate calculations.